File: app/src/main/java/me/zhanghai/android/files/provider/archive/ArchiveFileSystem.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive;

import android.os.Parcel;
import android.os.Parcelable;

import org.apache.commons.compress.archivers.ArchiveEntry;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;

import androidx.annotation.NonNull;
import java8.nio.file.Path;
import me.zhanghai.android.files.provider.common.ByteString;
import me.zhanghai.android.files.provider.common.ByteStringListPathFactory;
import me.zhanghai.android.files.provider.remote.RemoteFileSystemException;
import me.zhanghai.android.files.provider.root.RootableFileSystem;

class ArchiveFileSystem extends RootableFileSystem implements ByteStringListPathFactory {

    static final byte SEPARATOR = LocalArchiveFileSystem.SEPARATOR;

    @NonNull
    private final Path mArchiveFile;

    ArchiveFileSystem(@NonNull ArchiveFileSystemProvider provider, @NonNull Path archiveFile) {
        super(fileSystem -> new LocalArchiveFileSystem((ArchiveFileSystem) fileSystem, provider,
                archiveFile), RootArchiveFileSystem::new);

        mArchiveFile = archiveFile;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    protected LocalArchiveFileSystem getLocalFileSystem() {
        return super.getLocalFileSystem();
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    protected RootArchiveFileSystem getRootFileSystem() {
        return super.getRootFileSystem();
    }

    @NonNull
    ArchivePath getRootDirectory() {
        return getLocalFileSystem().getRootDirectory();
    }

    @NonNull
    ArchivePath getDefaultDirectory() {
        return getLocalFileSystem().getDefaultDirectory();
    }

    @NonNull
    Path getArchiveFile() {
        return getLocalFileSystem().getArchiveFile();
    }

    @NonNull
    ArchiveEntry getEntryAsLocal(@NonNull Path path) throws IOException {
        return getLocalFileSystem().getEntry(path);
    }

    @NonNull
    InputStream newInputStreamAsLocal(@NonNull Path file) throws IOException {
        return getLocalFileSystem().newInputStream(file);
    }

    @NonNull
    List<Path> getDirectoryChildrenAsLocal(@NonNull Path directory) throws IOException {
        return getLocalFileSystem().getDirectoryChildren(directory);
    }

    @NonNull
    String readSymbolicLinkAsLocal(@NonNull Path link) throws IOException {
        return getLocalFileSystem().readSymbolicLink(link);
    }

    void refresh() {
        getLocalFileSystem().refresh();
        getRootFileSystem().refresh();
    }

    void doRefreshIfNeededAsRoot() throws RemoteFileSystemException {
        getRootFileSystem().doRefreshIfNeeded();
    }

    @NonNull
    @Override
    public ArchivePath getPath(@NonNull ByteString first, @NonNull ByteString... more) {
        return getLocalFileSystem().getPath(first, more);
    }


    public static final Creator<ArchiveFileSystem> CREATOR = new Creator<ArchiveFileSystem>() {
        @Override
        public ArchiveFileSystem createFromParcel(Parcel source) {
            Path archiveFile = source.readParcelable(Path.class.getClassLoader());
            return ArchiveFileSystemProvider.getOrNewFileSystem(archiveFile);
        }
        @Override
        public ArchiveFileSystem[] newArray(int size) {
            return new ArchiveFileSystem[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable((Parcelable) mArchiveFile, flags);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/archive/ArchiveFileSystem.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive

import android.os.Parcel
import android.os.Parcelable
import java8.nio.file.Path
import me.zhanghai.android.files.provider.archive.archiver.ReadArchive
import me.zhanghai.android.files.provider.common.ByteString
import me.zhanghai.android.files.provider.common.ByteStringListPathCreator
import me.zhanghai.android.files.provider.remote.RemoteFileSystemException
import me.zhanghai.android.files.provider.root.RootableFileSystem
import java.io.IOException
import java.io.InputStream

internal class ArchiveFileSystem(
    provider: ArchiveFileSystemProvider,
    archiveFile: Path
) : RootableFileSystem(
    { LocalArchiveFileSystem(it as ArchiveFileSystem, provider, archiveFile) },
    { RootArchiveFileSystem(it) }
), ByteStringListPathCreator {
    override val localFileSystem: LocalArchiveFileSystem
        get() = super.localFileSystem as LocalArchiveFileSystem

    override val rootFileSystem: RootArchiveFileSystem
        get() = super.rootFileSystem as RootArchiveFileSystem

    val rootDirectory: ArchivePath
        get() = localFileSystem.rootDirectory

    val defaultDirectory: ArchivePath
        get() = localFileSystem.defaultDirectory

    val archiveFile: Path
        get() = localFileSystem.archiveFile

    @Throws(IOException::class)
    fun getEntryAsLocal(path: Path): ReadArchive.Entry = localFileSystem.getEntry(path)

    @Throws(IOException::class)
    fun newInputStreamAsLocal(file: Path): InputStream = localFileSystem.newInputStream(file)

    @Throws(IOException::class)
    fun getDirectoryChildrenAsLocal(directory: Path): List<Path> =
        localFileSystem.getDirectoryChildren(directory)

    @Throws(IOException::class)
    fun readSymbolicLinkAsLocal(link: Path): String = localFileSystem.readSymbolicLink(link)

    fun addPassword(password: String) {
        localFileSystem.addPassword(password)
        rootFileSystem.addPassword(password)
    }

    fun setPasswords(passwords: List<String>) {
        localFileSystem.setPasswords(passwords)
        rootFileSystem.setPasswords(passwords)
    }

    fun refresh() {
        localFileSystem.refresh()
        rootFileSystem.refresh()
    }

    @Throws(RemoteFileSystemException::class)
    fun prepareAsRoot() {
        rootFileSystem.prepare()
    }

    override fun getPath(first: ByteString, vararg more: ByteString): ArchivePath =
        localFileSystem.getPath(first, *more)

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(archiveFile as Parcelable, flags)
    }

    companion object {
        const val SEPARATOR: Byte = LocalArchiveFileSystem.SEPARATOR

        @JvmField
        val CREATOR = object : Parcelable.Creator<ArchiveFileSystem> {
            override fun createFromParcel(source: Parcel): ArchiveFileSystem {
                val archiveFile = source.readParcelable<Parcelable>(Path::class.java.classLoader)
                    as Path
                return ArchiveFileSystemProvider.getOrNewFileSystem(archiveFile)
            }

            override fun newArray(size: Int): Array<ArchiveFileSystem?> = arrayOfNulls(size)
        }
    }
}
